### PR TITLE
Developer guidelines (primarily navigating the pr and code review process)

### PR DESCRIPTION
**Pull request by jdonenine**
**Summary**

Publish an initial version of a set of "Developer Guidelines" that discuss the process, roles, and expectations of the pull request/code review process within the project.

**What issue does this change address and why do we need it?**

This adds documentation to the project aimed at aligning one of our core development practice (code reviews) more directly with our community values.  The focus here is on behaviors and expectations, not so much process. It aims to help make the pull request/code review process a more engaging, productive, efficient, and positive experience for those involved in it.

**How does this address the issue?**

It aims to create some behavioral context for the process as well as establish some guidelines and expected behaviors for those involved.

**How can this issue be reproduced?**

Post a PR 😉 

**How can this change be tested?**

Post a PR a few months from now 😄 

**Which specific GitHub issue(s) does this PR address?**:

All of them?

**Is there anything missing from this PR?**

This PR is currently only the guidelines portion of things, I believe that accompanying changes to our PR templates and perhaps other pieces of our developer docs should also be updated to reflect the new guidance and focus on the process aspect of things.  In fact I'm hoping that perhaps the structure of this PR itself could potentially become a new PR template, so please consider that as you review here and we'll address that next, separately. 

I am also a bit unsure of where this exactly should live, but I wanted this change to go through the PR process itself, for obvious reasons 😉  and I don't find our current community website page conducive to gathering feedback on changes.  So, for now, it goes into the repo and we can migrate the content wherever might make the most sense later.

**Checklist**
- [ ] Changes manually tested
- [ ] Automated Tests added/updated
- [x] Documentation added/updated
- [ ] CHANGELOG.md updated (not required for documentation PRs)
- [x] CLA Signed:  [DataStax CLA](https://cla.datastax.com/)

### CODE REVIEW

**Review by jeffbanks — APPROVED** [review]
Looks great overall.  I believe we are planning to branch a few of these ideas as separate threads for more detail.  This looks like a great foundation for developer guidelines that are impactful.

**Review by Miles-Garnsey — APPROVED** [review]
Approving.

File: dev-guidelines.md [hunk 1]
```diff
@@ -0,0 +1,194 @@
...
+
+Reviews are not free and should be planned for as we think about both our short and long term commitments both individually and as a project.
+
+**Learn** *[Beginner's mind, Value time]*
+
+Getting the most out of our time means learning from the observations and outcomes of the process.  It's the responsibility of the full team to learn from reviews and improve both themselves as well as the process itself.
+
+Remember that this process is iterative, we will repeat it many-many...many times in the lifetime of the project and each of those is a chance to do it in a better way.  When you find those better ways, come back here and update this guide :wink:.
+
+#### Author
```
[the remark this answers is not in the document]
Comment (jdonenine) [reply dev-guidelines.md:114]: In my latest commits I did end up including some of the content discussed here in these comments, particularly about assuming the best in others.

File: dev-guidelines.md [hunk 2]
```diff
@@ -0,0 +1,194 @@
...
+
+**Be Patient, Listen, Respond** *[Beginner's mind, Value time, Openness, Positivity]*
+
+Much like the same reponsiblity shared by the author, you the reviewer are expected to be patient with an author as they too juggle priorities to respond to your questions and feedback and then hear out their point of view when they do.  Support a two way conversation and keep in mind that you too can learn from this process.
+
+**Favor Approval Over Perfection** *[Value time, Positivity]*
+
+Having a bias towards forward progress doesn't have to sacrifice quality.  It does mean acknowledging that code is rarely perfect and does not have to be in order to push the project in a positive direction.
+
+***Reviewers should favor approval when a pull request has reached a state where it is a positive improvement on the overall codebase, even if there is room for further improvement.***  Reviewers should keep in mind that a contribution doesn't need to be done in the exact form and fashion that they would have done it to positively impact our project.
```
[the remark this answers is not in the document]
Comment (jdonenine) [reply dev-guidelines.md:164]: I followed up this conversation with a direct convo with @jsanda to get a better understanding of where he was coming from here.  My later commits add some content related to this conversation that I think help to bridge the gap we had here in it's initial form.  The overall message being, none of this is about relaxing our removing our standards, it's about doing a better job of how we go about upholding them.

**Comment by jdonenine** [comment]
I've gone ahead as discussed and resolved all of the discussion points that I felt like had been sorted out in discussion or that I addressed with my later changes over the last few days.

I've also created a separate issue and referenced a number of the remaining outstanding comments here in this doc that I felt like could be better addressed in the next phase of things, adding more process oriented content.  Although, in my last commit I did add a new section to begin that discussion.  It's not a "step by step" or anything, but I think it helps bridge the gap.

Here is the new issue for reference:  https://github.com/k8ssandra/k8ssandra/issues/1163

If anyone feels like points have been addressed either with the later changes or what will come in the new issues (hopefully in the next week or so) let me know, otherwise I think we're nearly ready to merge if the remaining few folks can all take another pass and approve if they are ready:  @adutra @burmanm @jsanda 

Thanks!  I know this was a lot to get through and not always the easiest topic to discuss openly!

**Pull request merged**